std-bits/base/src/main/java/org/enso/base/enso_cloud/TransientHTTPResponseCache.java

```diff
@@ -0,0 +1,428 @@
+      List<Pair<String, String>> resolvedHeaders,
+      String cacheKey,
+      RequestMaker requestMaker)
+      throws InterruptedException, IOException, ResponseTooLargeException {
+    assert !cache.containsKey(cacheKey);
+
+    var response = requestMaker.run();
+
+    // Don't cache non-200 repsonses.
+    if (response.statusCode() != 200) {
+      return response;
+    }
+
+    // If we get a content-length header, make room for the new file before
+    // downloading. If the response is too large, throw
+    // ResponseTooLargeException.
+    var sizeMaybe = response.headers().firstValue("content-length").map(Long::parseLong);
+    if (sizeMaybe.isPresent()) {
+      long size = sizeMaybe.get();
+      if (size > getMaxFileSize()) {
+        throw new ResponseTooLargeException(
+            size, getMaxFileSize(), unresolvedURI.render().toString());
+      }
+      makeRoomFor(size);
+    }
+
+    try {
+      // Download the response data.
+      int ttl = calculateTTL(response.headers());
+      ZonedDateTime expiry = getNow().plus(Duration.ofSeconds(ttl));
+      int statusCode = response.statusCode();
+      var headers = response.headers();
+      String responseDataPath = downloadResponseData(unresolvedURI, response);
+      long size = new File(responseDataPath).length();
+
+      // Create a cache entry. Remove old files if we are over the cache size
+      // limit.
+      var cacheEntry =
+          new CacheEntry(resolvedURI, headers, responseDataPath, size, statusCode, expiry);
+      cache.put(cacheKey, cacheEntry);
+      markCacheEntryUsed(cacheKey);
+
+      // Clear out old entries to satisfy the total cache size limit. This might
+      // be necessary here if we didn't receive a content-length.
+      removeFilesToSatisfyLimit();
+
+      return buildEnsoHttpResponseFromCacheEntry(cacheEntry);
+    } catch (IOException e) {
+      logger.log(
+          Level.WARNING,
+          "Failure when caching HTTP response, will re-issue request without caching: " + e);
+      // Re-issue the request since we don't know if we've consumed any of the response.
+      return requestMaker.run();
+    }
+  }
+
+  /** Create an EnsoHttpResponse with an InputStream reading from the cache file. */
+  private static EnsoHttpResponse buildEnsoHttpResponseFromCacheEntry(CacheEntry cacheEntry)
+      throws IOException {
+    InputStream inputStream = new FileInputStream(cacheEntry.responseDataPath);
+    return new EnsoHttpResponse(
+        cacheEntry.uri(), cacheEntry.headers(), inputStream, cacheEntry.statusCode());
+  }
+
+  /**
+   * Read the repsonse data from the remote server into the cache file. If the downloaded data is
+   * over the file size limit, throw a ResponseTooLargeException.
+   */
+  private static String downloadResponseData(
+      URIWithSecrets unresolvedURI, EnsoHttpResponse response)
+      throws IOException, ResponseTooLargeException {
+    File temp = File.createTempFile("TransientHTTPResponseCache", "");
```

Comment:
Worth putting the hash into the file name?

Reply:
Done